[Serializer] Add support for collecting type error during denormalization

lyrixx opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 5.4
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | Fix #27824, Fix #42236, Fix #38472, Fix #37419 Fix #38968
| License       | MIT
| Doc PR        |

---

There is something that I don't like about the (de)Serializer. It's about the way it deals with typed properties.
As soon as you add a type to a property, the API can return 500.

Let's consider the following code:
```php
class MyDto
{
    public string $string;
    public int $int;
    public float $float;
    public bool $bool;
    public \DateTime $dateTime;
    public \DateTimeImmutable $dateTimeImmutable;
    public \DateTimeZone $dateTimeZone;
    public \SplFileInfo $splFileInfo;
    public Uuid $uuid;
    public array $array;
    /** @var MyDto[] */
    public array $collection;
}
```

and the following JSON:

```json
{
	"string": null,
	"int": null,
	"float": null,
	"bool": null,
	"dateTime": null,
	"dateTimeImmutable": null,
	"dateTimeZone": null,
	"splFileInfo": null,
	"uuid": null,
	"array": null,
	"collection": [
		{
			"string": "string"
		},
		{
			"string": null
		}
	]
}
```

**By default**, I got a 500:

<details>

![image](https://user-images.githubusercontent.com/408368/129211588-0ce9064e-171d-42f2-89ac-b126fc3f9eab.png)

</details>

It's the same with the prod environment. This is far from perfect when you try to make a public API :/ 
ATM, the only solution, is to remove all typehints and add assertions (validator component). With that, the public API is nice, but the internal PHP is not so good (PHP 7.4+ FTW!)

In APIP, they have support for transforming to [something](https://github.com/api-platform/core/blob/53837eee3ebdea861ffc1c9c7f052eecca114757/src/Core/Serializer/AbstractItemNormalizer.php#L233-L237) they can handle gracefully. But the deserialization stop on the first error (so the end user must fix the error, try again, fix the second error, try again etc.). And the raw exception message is leaked to the end user. So the API can return something like  `The type of the "string" attribute for class "App\Dto\MyDto" must be one of "string" ("null" given).`. Really not cool :/ 

So ATM, building a nice public API is not cool.

That's why I propose this PR that address all issues reported
* be able to collect all error
* with their property path associated
* don't leak anymore internal 

In order to not break the BC, I had to use some fancy code to make it work 🐒

With the following code, I'm able to collect all errors, transform them in `ConstraintViolationList` and render them properly, as expected.

<details>

![image](https://user-images.githubusercontent.com/408368/129215560-b0254a4e-fec7-4422-bee0-95cf9f9eda6c.png)

</details>

```php
    #[Route('/api', methods:['POST'])]
    public function apiPost(SerializerInterface $serializer, Request $request): Response
    {
        try {
           $dto = $serializer->deserialize($request->getContent(), MyDto::class, 'json', [
                DenormalizerInterface::COLLECT_DENORMALIZATION_ERRORS => true,
            ]);
        } catch (PartialDenormalizationException $e) {
            $violations = new ConstraintViolationList();
            /** @var NotNormalizableValueException */
            foreach ($e->getErrors() as $exception) {
                $message = sprintf('The type must be one of "%s" ("%s" given).', implode(', ', $exception->getExpectedTypes()), $exception->getCurrentType());
                $parameters = [];
                if ($exception->canUseMessageForUser()) {
                    $parameters['hint'] = $exception->getMessage();
                }
                $violations->add(new ConstraintViolation($message, '', $parameters, null, $exception->getPath(), null));
            };

            return $this->json($violations, 400);

        }

        return $this->json($dto);
    }
```

If this PR got accepted, the above code could be transferred to APIP to handle correctly the deserialization

lyrixx committed:
[Serializer] Add support for collecting type error during denormalization

jderusse commented on src/Symfony/Component/Serializer/Normalizer/AbstractObjectNormalizer.php:
```diff
@@ -375,12 +378,33 @@ public function denormalize($data, string $type, string $format = null, array $c
+                try {
+                    $value = $this->validateAndDenormalize($types, $resolvedClass, $attribute, $value, $format, $attributeContext);
+                } catch (NotNormalizableValueException $exception) {
+                    if (null === ($context['not_normalizable_value_exceptions'] ?? null)) {
```
`if (!isset(...))`?
same bellow

jderusse reviewed (commented):
replaces https://github.com/symfony/symfony/issues/27824 ?

I'm mix feeling with this PR that brings validation without to much changes. But also silents deserialization failures..

Kocal commented on src/Symfony/Component/Serializer/Exception/NotNormalizableValueException.php:
```diff
@@ -16,4 +16,46 @@
+    {
+        $self = new self($message, $code, $previous);
+
+        $currentType = get_debug_type($data);
+        $self->currentType = $currentType;
```
```suggestion
        $self->currentType = get_debug_type($data);
```

Kocal commented on src/Symfony/Component/Serializer/Tests/SerializerTest.php:
```diff
@@ -713,6 +719,164 @@ public function testDeserializeAndUnwrap()
+                new UidNormalizer(),
+                new ObjectNormalizer($classMetadataFactory, null, null, $extractor, new ClassDiscriminatorFromClassMetadata($classMetadataFactory))
+            ],
+            ['json' => new JsonEncoder()
+        ]);
```
```suggestion
            ['json' => new JsonEncoder()]
        );
```

lyrixx commented:
> I'm mix feeling with this PR that brings validation without to much changes. But also silents deserialization failures..

I understand this is a bit weird. But this process has an advantage overs others PR: you can get the hydrated object **and** the errors.
So you can merge deserialization errors with validations errors and return everything to the end user.

jderusse commented:
> > I'm mix feeling with this PR that brings validation without to much changes. But also silents deserialization failures..
> 
> I understand this is a bit weird. But this process has an advantage overs others PR: you can get the hydrated object **and** the errors.
> So you can merge deserialization errors with validations errors and return everything to the end user.

Yeah, that's what I like in your PR!.

What's about
```
try {
   $o = $serializer->unserialize(...);
} catch (PartialDenormalizationException $e) {
  $o = $e->getPartialObject();
  $exceptions = $e->getExceptions();
}
```

lyrixx commented:
Oh, this is a good idea too! I'll update the PR

ro0NL commented on src/Symfony/Component/Serializer/Exception/NotNormalizableValueException.php:
```diff
@@ -16,4 +16,46 @@
+    private $useMessageForUser = null;
+
+    /**
+     * @param bool $useMessageForUser If the message passed to this exception is something that can be shown
```
WDYT of `InvalidTypeException/InvalidValueException extends NotNormalizableValueException`? i tend to believe it makes semantics a bit more straightforward :)

lyrixx replied:
Why not, but this is not related to the useMessageForUser right ?

ro0NL replied:
IIUC currently any useMessageForUser=true case, is actually a InvalidTypeException/InvalidValueException case

reasonably i'd say

catch(InvalidTypeException/InvalidValueException) => 400 + user message (eg. useMessageForUser=true)
catch(NotNormalizableValueException) => 400 + safe message (eg. useMessageForUser=false)

ro0NL replied:
the caught exception could be a compound one, to further digest. Or as you obtain them currenly, by ref. Then the context could be determined using instanceof.

ro0NL commented:
about partial objects, im not sure they should ever leak.

> you can get the hydrated object and the errors

dont see the usecase yet :thinking:

lyrixx commented:
>about partial objects, im not sure they should ever leak.

 I'm not sure I follow you :/ 

> dont see the usecase yet

See the PR  description => you can get the whole (type + validator) validation at once

ro0NL commented:
```
} catch (PartialDenormalizationException $e) {
  $o = $e->getPartialObject();
```

would be leaking something invalid isnt it?

IIUC in your PR example code you dont actually need a partial `$dto` within the `if ($exceptions) {` block. Or why would you?

lyrixx commented:
> would be leaking something invalid isnt it?

anyway, it's always the case, that's why we validate the data after. Same with form (but the form component trigger the validation itself) 



> IIUC in your PR example code you dont actually need a partial `$dto` within the `if ($exceptions) {` block. Or why would you?

In my example, it don't trigger the validation, but it would be even better if it's the case ❤️

ro0NL commented:
currently no invalid object leaks isnt it? you get a NotNormalizableValueException

then i agree with @jderusse , as i have mixed feelings about silencing those now.

ro0NL commented:
> In my example, it don't trigger the validation, but it would be even better if it's the case 

well i think we should serialize the NotNormalizableValueException to a constration violation list yes

ro0NL commented:
ohh =) you mean `$validator->validate($dto)`, after deserialize :facepalm: i see we have 2 point of views.

im really not sure one should continue from an invalid state, assuming validator does some job maybe.

lyrixx commented:
> im really not sure one should continue from an invalid state, assuming validator does some job maybe.

It's up to the final developper, but why not?

```
class Test
{
    public int $int; // required, but a string is passed
    public bool $bool; // required, nothing is passed
}
```

* the "int" value must be a int, string passed
* the "bool" value must not be blank

=> this is nice for the end user to report all errors. But again, it's up to the developer to choose

ro0NL commented:
got ya. Reporting deserialization AND validation in one go is also nice yes :+1: (overread https://github.com/symfony/symfony/pull/42502#issuecomment-899370779)

I agree with https://github.com/symfony/symfony/pull/42502#issuecomment-899405106 to optin :+1: 

sorry if i was noisy :)

lyrixx commented:
I pushed another implementation where I introduced a `PartialDenormalizationException` as asked by @jderusse 

I also addressed CS comments.

I'm not sure about the @ro0NL comment to add other exception classes. What do other people think ?

jderusse commented on src/Symfony/Component/Serializer/Normalizer/DenormalizerInterface.php:
```diff
@@ -24,13 +24,15 @@
      * @param string $type    The expected class to instantiate
      * @param string $format  Format the given data was extracted from
-     * @param array  $context Options available to the denormalizer
+     * @param array  $context options available to the denormalizer
```
?

lyrixx replied:
php cs fixer, I'll revert that

ro0NL commented on src/Symfony/Component/Serializer/Normalizer/DateTimeNormalizer.php:
```diff
@@ -86,7 +87,7 @@ public function denormalize($data, string $type, string $format = null, array $c
 
         if (null === $data || (\is_string($data) && '' === trim($data))) {
-            throw new NotNormalizableValueException('The data is either an empty string or null, you should pass a string that can be parsed with the passed format or a valid DateTime string.');
+            throw NotNormalizableValueException::create('The data is either an empty string or null, you should pass a string that can be parsed with the passed format or a valid DateTime string.', $data, [Type::BUILTIN_TYPE_STRING], $context['deserialization_path'] ?? null, true);
```
below we call this `datetime string`, here we leak a symbol name IIUC

lyrixx replied:
IMHO, This is not something sensitive. It only leak the app is written in PHP but it give a real hint for the end user. 
WDYT?

ro0NL commented on src/Symfony/Component/Serializer/Normalizer/DateTimeNormalizer.php:
```diff
@@ -98,13 +99,13 @@ public function denormalize($data, string $type, string $format = null, array $c
             return \DateTime::class === $type ? new \DateTime($data, $timezone) : new \DateTimeImmutable($data, $timezone);
         } catch (\Exception $e) {
-            throw new NotNormalizableValueException($e->getMessage(), $e->getCode(), $e);
+            throw NotNormalizableValueException::create($e->getMessage(), $data, [Type::BUILTIN_TYPE_STRING], $context['deserialization_path'] ?? null, true, $e->getCode(), $e);
```
is this an internal message leaking? :see_no_evil:

ro0NL commented on src/Symfony/Component/Serializer/Normalizer/UidNormalizer.php:
```diff
@@ -72,7 +73,7 @@ public function denormalize($data, string $type, string $format = null, array $c
             return Ulid::class === $type ? Ulid::fromString($data) : Uuid::fromString($data);
         } catch (\InvalidArgumentException $exception) {
-            throw new NotNormalizableValueException(sprintf('The data is not a valid "%s" string representation.', $type));
+            throw NotNormalizableValueException::create(sprintf('The data is not a valid "%s" string representation.', $type), $data, [Type::BUILTIN_TYPE_STRING], $context['deserialization_path'] ?? null, true);
```
also, $type is leaking a SF class (but i like this message generally)

lyrixx replied:
IMHO, This is not something sensitive. It only leak the app is written in PHP but it give a real hint for the end user.
WDYT?

ro0NL replied:
generally i'd say we talk "a valid UUID string", similar like we talk about a "datetime string"

lyrixx replied:
You were right, I changed this code and added another catch (type error, If I pass `null`)

ro0NL commented on src/Symfony/Component/Serializer/Serializer.php:
```diff
@@ -204,12 +205,16 @@ public function normalize($data, string $format = null, array $context = [])
      */
     public function denormalize($data, string $type, string $format = null, array $context = [])
     {
+        if (isset($context[DenormalizerInterface::COLLECT_DENORMALIZATION_ERRORS], $context['not_normalizable_value_exceptions'])) {
```
`if (isset($context['not_normalizable_value_exceptions'])) {` ?

lyrixx replied:
This is not possible, since the serializer is called recursively

ro0NL reviewed (commented):
i withdraw from my previous InvalidTypeException suggestion. Having the boolean flag on NotNormalizableValueException is likely easier to discover :+1: and covers all our usecases.

i find NotNormalizableValueException vs UnexpectedValueException generally confusing. Is eg. DateIntervalNormalizer relevant?

im not sure we verified all caught exception messages :) i believe this is the point where we should transform them to be more friendly, eg. UidNormalizer like.

Im not sure `$parameters['hint'] = $exception->getMessage();` is useful in terms of i18n. IMHO the error is either a type related or a value related one.

lyrixx commented:
* I fixed test for PHP 7.2
* I addressed some @ro0NL comments
* I added support collecting error that occurs during the object instanciation 
* I fixed a type error in the UidNormalizer

ro0NL commented:
@lyrixx LGTM. I cannot decide on further Serializer semantics (in terms of i18n / exposing PHP-SF symbols). I believe it doesn't matter much _yet_, for what i think is the key feature in this PR: "use sf/validator to validate a partial deserialized object from sf/serializer".

Using sf/validator solves my issues generally if we talk about eg. human friendly errors. My other goal would be to enforce fully deserialized objects OR human friendly errors from _just_ sf/serializer. But we pursue similar goals :+1:

BafS commented on src/Symfony/Component/Serializer/Exception/NotNormalizableValueException.php:
```diff
@@ -16,4 +16,45 @@
+        return $this->currentType;
+    }
+
+    public function getExpectedTypes(): ?array
```
what about adding more type information when possible:
```suggestion
    /**
     * @return string[]|null
     */
    public function getExpectedTypes(): ?array
```

lyrixx replied:
Fixed, thanks

fabpot commented on src/Symfony/Component/Serializer/CHANGELOG.md:
```diff
@@ -7,6 +7,7 @@ CHANGELOG
  * Add support of PHP backed enumerations
  * Add support for serializing empty array as object
  * Deprecate support for returning empty, iterable, countable, raw object when normalizing
+ * Add support for collecting type error during denormalization
```
```suggestion
 * Add support for collecting type errors during denormalization
```

fabpot commented on src/Symfony/Component/Serializer/Exception/NotNormalizableValueException.php:
```diff
@@ -16,4 +16,45 @@
+    private $currentType = null;
+    private $expectedTypes = null;
+    private $path = null;
+    private $useMessageForUser = null;
```
The `= null` parts can be removed, right?

lyrixx replied:
Yes, but this was in preparation for PHP 7.4 typed properties

fabpot commented on src/Symfony/Component/Serializer/Exception/NotNormalizableValueException.php:
```diff
@@ -16,4 +16,45 @@
+     *                                safely to your user. In other words, avoid catching other exceptions and
+     *                                passing their message directly to this class.
+     */
+    public static function create(string $message, $data, array $expectedTypes, string $path = null, bool $useMessageForUser = false, int $code = 0, \Throwable $previous = null): self
```
We are not doing this anywhere else (hope I'm right here), so I would prefer a proper `__construct` method here if possible.

lyrixx replied:
This would break the BC?
And if I keep the BC, then the code to create the exception will be ugly.
Wdyt ?

lyrixx replied:
what should I do ? /cc  @fabpot

fabpot replied:
Keeping BC means supporting both the old and new signature, which is something we do all the time, right? Or am I missing something?

lyrixx replied:
you are right, but it means I will have to write such code

```
throw new NotNormalizableValueException($message, 0, null, $data, $expectedTypes, $path, $useMessageForUser)
```

(I don't like the `0, null`)

fabpot replied:
So, does it mean that both the constructor and the new create methods are still be used? If that the case, we need a more descriptive name than create.

lyrixx replied:
IIRC the regular way is not used anymore in Symfony, but I cannot say it's not used in userland application :/ 

I'll try to find a better name 👍🏼

lyrixx replied:
Here we go, I hope it's better

lyrixx commented:
Hi @fabpot. I fixed your comments and there is already 2 approvals

fabpot commented:
Thank you @lyrixx.

julienfalque commented:
Thanks @lyrixx!

roman-eonx commented:
Thanks, @lyrixx for this fix we were waiting that long! Are you planning to make a PR into APIP to start using this new feature?
